### PR TITLE
Add KMS functionality

### PR DESCRIPTION
**Pull request by torao**
## Description

This PR enables the use of a secure KMS (Key Management System) from Ostracon.

Currently, the key pairs used by Ostracon for agreements are stored locally. But Ostracon nodes are at risk of having their private keys stolen when a malicious attacker breaks into the system since they must accept inbound connections from the Internet.

For more mission-critical applications, such as finance, the node storing the private key must be independent of the Ostracon agreement node, placed behind a firewall, and able to connect to solutions using an HSM (Hardware Security Module) or our SGX solution.

Tendermint, a fork of Ostracon, has the ability to connect to KMS using [`tmkms`](https://github.com/iqlusioninc/tmkms) communication protocol. Note, however, that the `init` and `unsafe-reset-priv-validator` subcommands of the Tendermint CLI are *not* designed to configure KMS, and this PR *still* requires copying the public key stored in the KMS to a local key file. Thus, you should be set up as follows if you use your KMS:

1. Create an initial configuration with the `ostracon init` command.
2. In `~/.ostracon/config/priv_validator_key.json`, replace `pubkey/value` field with the Base64 representation of the public key stored in KMS.
3. In `~/.ostracon/config/genesis.json`, replace the `validators/address` and `validators/pubkey/value` fields with the values from the KSM public key in the same way if the node is the initial validator.

This PR contains the following changes:

* Connection test with KSM at the library level. This is a program that starts with `main()` instead of a unit test.
* `ostracon run`: Removed the behavior of referencing local key pair when `priv_validator_laddr` is set.
* `ostracon show-validator`: Changed to get and display KSM public key when `priv_validator_laddr` is set.

### CODE REVIEW

File: cmd/ostracon/commands/kms_test.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	"testing"
+	"time"
+
+	"github.com/line/ostracon/crypto"
+	"github.com/line/ostracon/crypto/ed25519"
+	"github.com/line/ostracon/libs/log"
+	"github.com/line/ostracon/privval"
+)
+
+func WithMockKMS(t *testing.T, dir, chainID string, f func(string, crypto.PrivKey)) {
```
Comment (tnasu) [inline cmd/ostracon/commands/kms_test.go:17]: Should move into `privval` or `node` package.

**Review by tnasu — CHANGES_REQUESTED** [review]

File: cmd/ostracon/commands/show_validator.go [hunk 2]
```diff
@@ -15,18 +18,29 @@ var ShowValidatorCmd = &cobra.Command{
 	Use:     "show-validator",
 	Aliases: []string{"show_validator"},
 	Short:   "Show this node's validator info",
-	RunE:    showValidator,
-	PreRun:  deprecateSnakeCase,
+	RunE: func(cmd *cobra.Command, args []string) error {
```
Comment (kokeshiM0chi) [inline cmd/ostracon/commands/show_validator.go:21]: Do you use the value specified in config to decide whether to return information on kms node or not?

File: cmd/ostracon/commands/init.go [hunk 3]
```diff
@@ -73,6 +73,10 @@ func initFilesWithConfig(config *cfg.Config) error {
 		logger.Info("Found genesis file", "path", genFile)
 	} else {
 		genDoc := types.GenesisDoc{
+			// FIXME: It shouldn't be generated automatically, and the config and genDoc chainIDs should be the same.
```
[the remark this answers is not in the document]
Comment (torao) [reply cmd/ostracon/commands/init.go:76]: I had mistakenly thought that the `BaseConfig.chainID` field was a duplicate setting also described in `config/config.toml` independent of `genDoc`, even though it has the same semantics. I've confirmed that this field is not subject to serialization/deserialization because it doesn't have a `mapstructure` and is, therefore, volatile information rather than persistent information stored in `config.toml`. Thus, I'll remove this comment.

Upon my exam, it appears that `BaseConfig.chainID` isn't used outside of this location and tests. I guess we need to discuss whether it's really a necessary field, but that should be done outside of this PR.

File: cmd/ostracon/commands/run_node_test.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+	"github.com/line/ostracon/crypto"
+	nm "github.com/line/ostracon/node"
+	"github.com/line/ostracon/types"
+	"github.com/stretchr/testify/require"
+)
+
+// RunCmd.RunE can only be stopped by a signal for process. For this reason, NewOstraconNode() is called directly
+// to test.
+func TestNewOstraconNode(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (torao) [reply cmd/ostracon/commands/run_node_test.go:16]: The tests added to this file are almost identical to those added to `node_test.go` (for the reasons given in the comments of this function). I thought a bit about whether there was anything I could test against `RunCmd`, but it seems unlikely. I'll delete this file from PR.

File: cmd/ostracon/commands/run_node_test.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	require.NoError(t, err)
+
+	// verify the retrieved public key matches the local public key
+	expected := loadFilePVKey(t, config.PrivValidatorKeyFile()).PubKey
+	actual, err := node.PrivValidator().GetPubKey()
+	require.NoError(t, err)
+	require.Equal(t, expected, actual)
+}
+
+func TestNewOstraconNodeWithKMS(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (torao) [reply cmd/ostracon/commands/run_node_test.go:45]: ditto.

File: cmd/ostracon/commands/show_validator.go [hunk 6]
```diff
@@ -15,18 +18,29 @@ var ShowValidatorCmd = &cobra.Command{
...
+		return showValidator(cmd, args, config)
+	},
+	PreRun: deprecateSnakeCase,
 }
 
-func showValidator(cmd *cobra.Command, args []string) error {
-	keyFilePath := config.PrivValidatorKeyFile()
-	if !tmos.FileExists(keyFilePath) {
-		return fmt.Errorf("private validator file %s does not exist", keyFilePath)
+func showValidator(cmd *cobra.Command, args []string, config *cfg.Config) error {
```
[the remark this answers is not in the document]
Comment (torao) [reply cmd/ostracon/commands/show_validator.go:27]: Referencing the global variable `root.config` directly is inadvisable because it sometimes causes race conditions when tests are run. By passing it as a parameter, it can be made to work with independent configurations for each test case. Although this way is a test-driven requirement, I think this is a better way from the perspective of the degree of coupling.

File: cmd/ostracon/commands/show_validator.go (hunk 2 above)
Comment (torao) [reply cmd/ostracon/commands/show_validator.go:21]: To the first sentence, yes, I do. And what's the relationship between the first and second sentence of your point? Are you implying that this anonymous function should be given an easy-to-understand name? Even if I do that, I don't see how it would improve anything.

File: cmd/ostracon/commands/show_validator.go [hunk 7]
```diff
@@ -15,18 +18,29 @@ var ShowValidatorCmd = &cobra.Command{
...
+	var pv types.PrivValidator
+	var err error
```
[the remark this answers is not in the document]
Comment (torao) [reply cmd/ostracon/commands/show_validator.go:29]: It's possible but results in one more redundant line as follows, since `pv` is declared outside of `if`, so using `:=` at line 32 will result in a new `pv` with different scope.

```go
31: p, err := node.CreateAndStartPrivValidatorSocketClient(config.PrivValidatorListenAddr, chainID, logger)
...
35: pv = p
```

File: cmd/ostracon/commands/show_validator.go (hunk 2 above)
Comment (kokeshiM0chi) [reply cmd/ostracon/commands/show_validator.go:21]: Sorry for lack of the information. It was not an anonymous function, but the part of the processing after `if config.PrivValidatorListenAddr! =" "{` Inside `showValidator`. Having a function that clearly states that it is a Kms node helps to read the intent.

File: cmd/ostracon/commands/show_validator.go [hunk 8]
```diff
@@ -15,18 +18,29 @@ var ShowValidatorCmd = &cobra.Command{
...
+	if config.PrivValidatorListenAddr != "" {
+		chainID := "" // currently not in use
+		pv, err = node.CreateAndStartPrivValidatorSocketClient(config.PrivValidatorListenAddr, chainID, logger)
```
[the remark this answers is not in the document]
Comment (torao) [reply cmd/ostracon/commands/show_validator.go:32]: `config.ChainID()` isn't set anywhere except in testing, and I thought to avoid it because `chainID := config.ChainID()` could be a confusing bug that looks like correct code but is never actually the correct value.

Currently, neither ostracon-kms nor tmkms use `chainID` to identify the `PrivValidator`, so I assumed `""`, but I'll rewrite it to use the correct `chainID` according to [node.go(759)](https://github.com/line/ostracon/blob/7cb986b6492f480847caa2bab317d020da60d6ca/node/node.go#L759)'s method.

File: config/config.go [hunk 9]
```diff
@@ -9,8 +9,9 @@ import (
 	"path/filepath"
 	"time"
 
-	"github.com/line/ostracon/privval"
 	"github.com/line/tm-db/v2/metadb"
+
+	"github.com/line/ostracon/privval"
```
[the remark this answers is not in the document]
Comment (torao) [reply config/config.go:14]: This isn't the purpose of this fix, so I'll revert it.

This is the result of modifying `config.go` and running `make format`. The `make format` uses `goimports` with the `-local github.com/line/ostracon` option, and I think it's better to use it in CI as well.

File: config/toml.go [hunk 10]
```diff
@@ -575,6 +576,7 @@ func ResetTestRootWithChainID(testName string, chainID string) *Config {
 	// we always overwrite the priv val
 	tmos.MustWriteFile(privKeyFilePath, []byte(testPrivValidatorKey), 0644)
 	tmos.MustWriteFile(privStateFilePath, []byte(testPrivValidatorState), 0644)
+	tmos.MustWriteFile(nodeKeyFilePath, []byte(testNodeKey), 0644)
```
[the remark this answers is not in the document]
Comment (torao) [reply config/toml.go:579]: `ResetTestRoot()` is a common function to create the initial configuration of the test directory (I've confirmed it's called only from unit tests).

This time, I needed to generate the node key along with other files to be used for communication encryption to KMS.

File: node/node_test.go [hunk 11]
```diff
@@ -36,6 +36,25 @@ import (
...
+	defer os.RemoveAll(config.RootDir)
+	require.Equal(t, config.PrivValidatorListenAddr, "")
+	node, err := NewOstraconNode(config, log.TestingLogger())
+	require.NoError(t, err)
+	pubKey, err := node.PrivValidator().GetPubKey()
+	require.NoError(t, err)
+	require.NotNil(t, pubKey)
+}
+
+func TestNewOstraconNode_WithoutNodeKye(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (torao) [reply node/node_test.go:50]: oops. I'll fix it.

File: cmd/ostracon/commands/kms_test.go (hunk 1 above)
Comment (torao) [reply cmd/ostracon/commands/kms_test.go:17]: This file has been moved to `privval/test_util.go` because Golang’s test builds couldn't reference common functions defined in `*_test.go` outside the package.

Note that `utils_test.go` also exists in the same directory, Here is a test case for `utils.go`.

File: cmd/ostracon/commands/show_validator.go (hunk 2 above)
Comment (torao) [reply cmd/ostracon/commands/show_validator.go:21]: At this point, I don't think the code is so difficult that it's necessary to separate functions with and without `config.PrivValidatorListenAddr` for semantics and ease of testing. We may do so in future revisions if the if-else statements become more complex.

File: cmd/ostracon/commands/show_validator.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (tnasu) [reply cmd/ostracon/commands/show_validator.go:27]: Got it.

File: node/node_test.go [hunk 12]
```diff
@@ -9,15 +9,13 @@ import (
 	"testing"
 	"time"
 
+	dbm "github.com/line/tm-db/v2"
```
Comment (tnasu) [inline node/node_test.go:12]: Unnecessary changes?

File: node/node_test.go [hunk 13]
```diff
@@ -9,15 +9,13 @@ import (
...
+	"github.com/line/tm-db/v2/memdb"
```
Comment (tnasu) [inline node/node_test.go:13]: Unnecessary changes?

File: cmd/ostracon/commands/show_validator_test.go [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+	}()
+
+	config = cfg.DefaultConfig()
+	config.SetRoot(dir)
+	err := RootCmd.PersistentPreRunE(RootCmd, nil)
+	require.NoError(t, err)
+	init := NewInitCmd()
+	err = init.RunE(init, nil)
+	require.NoError(t, err)
+	require.FileExists(t, config.GenesisFile())
```
Comment (tnasu) [inline cmd/ostracon/commands/show_validator_test.go:140]: Does `loadChainID` depend on `RootCmd.PersistentPreRunE` and `init.RunE`?

I think `loadChainID` is a local function and it doesn't need to depend on some public commands.

If testing `loadChainID`, I think the below is uncessarry.
```
	config = cfg.DefaultConfig()
	config.SetRoot(dir)
	err := RootCmd.PersistentPreRunE(RootCmd, nil)
	require.NoError(t, err)
	init := NewInitCmd()
	err = init.RunE(init, nil)
	require.NoError(t, err)
	require.FileExists(t, config.GenesisFile())
```

File: cmd/ostracon/commands/show_validator_test.go [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+
+func TestLoadChainIDWithoutStateDB(t *testing.T) {
+	dir := setupEnv(t)
+	config := cfg.DefaultConfig()
+	config.SetRoot(dir)
+	err := RootCmd.PersistentPreRunE(RootCmd, nil)
+	require.NoError(t, err)
+	init := NewInitCmd()
+	err = init.RunE(init, nil)
+	require.NoError(t, err)
```
Comment (tnasu) [inline cmd/ostracon/commands/show_validator_test.go:159]: ditto

File: cmd/ostracon/commands/show_validator_test.go (hunk 14 above)
Comment (torao) [reply cmd/ostracon/commands/show_validator_test.go:131]: In order to use `loadChainID()`, we need an environment where genesis.json and state db exist. To create them formally, I thought it would be reliable and easy to create an initialized directory with the same action as `init`.

On the other hand, there are several ways to configure the initial directory for testing, and  `cfg.ResetTestRootWithChainID()` is one of them. I don't have a strong preference for the way the directory is initially configured for the test, so I'll modify this to do it that way.

File: node/node_test.go (hunk 12 above)
Comment (torao) [reply node/node_test.go]: It's an effect of the `make format`. Currently, imports in many source files are disjointed, separated by blank lines in meaningless units.

I'll fix this, and I think if the opportunity to correct them in an individual PR isn't given, we should organize them correctly somewhere and make the `make format` and CI behavior the same (`Makefile` seem better than CI at the moment), as a "good first issue".

```
gofmt -w -s
goimports -w -local github.com/line/ostracon
```

**Review by tnasu — APPROVED** [review]

File: cmd/ostracon/commands/show_validator.go (hunk 2 above)
Comment (kokeshiM0chi) [reply cmd/ostracon/commands/show_validator.go:21]: It's easy for those who know it, but it's not easy for those who don't.  I approve in the hope that the documentation will be enriched.

**Review by kokeshiM0chi — APPROVED** [review]
